pyomo/core/expr/numvalue.py

```diff
@@ -280,78 +257,65 @@ def polynomial_degree(obj):
     """
     if obj.__class__ in native_types:
         return 0
-    try:
-        return obj.polynomial_degree()
-    except AttributeError:
-        pass
-    return as_numeric(obj).polynomial_degree()
+    return obj.polynomial_degree()
 
 # It is very common to have only a few constants in a model, but those
 # constants get repeated many times.  KnownConstants lets us re-use /
 # share constants we have seen before.
-KnownConstants = {}
-
-def update_KnownConstants(obj, val):
-    if len(KnownConstants) < 100:
-        KnownConstants[obj] = val
+_KnownConstants = {}
 
 
 def as_numeric(obj):
```

Comment:
This PR fundamentally changes the meaning of `as_numeric()`.  Since pre-coopr 3.x, `as_numeric` returned its argument cast into a class *derived* from `NumericValue`.  The other key thing that `as_numeric` did was maintain / update the `native_numeric_types` set when users used types that Pyomo doesn't know about but behave like numeric types.  This PR changes that so that it is not clear what / how the `native_numeric_types` set gets updated.